src/database-postgresql.cpp

```diff
@@ -0,0 +1,234 @@
+{
+	if (PQping(m_connect_string.c_str()) != PQPING_OK) {
+		throw DatabaseException(std::string(
+			"PostgreSQL database error: ") +
+			PQerrorMessage(m_conn));
+	}
+}
+
+bool Database_PostgreSQL::initialized() const
+{
+	return (PQstatus(m_conn) == CONNECTION_OK);
+}
+
+void Database_PostgreSQL::initStatements()
+{
+	prepareStatement("read_block",
+			"SELECT block_data FROM blocks "
+					"WHERE posX = $1::int4 AND posY = $2::int4 AND posZ = $3::int4");
+
+	prepareStatement("write_block",
+			"INSERT INTO blocks (posX, posY, posZ, block_data) VALUES "
+					"($1::int4, $2::int4, $3::int4, $4::bytea) "
+					"ON CONFLICT ON CONSTRAINT blocks_pkey DO UPDATE SET block_data = $4::bytea");
+
+	prepareStatement("delete_block", "DELETE FROM blocks WHERE posX = $1::int4 AND posY = $2::int4 AND posZ = $3::int4");
+
+	prepareStatement("list_all_loadable_blocks", "SELECT posX, posY, posZ FROM blocks");
+}
+
+void Database_PostgreSQL::prepareStatement(const std::string &name, const std::string &sql)
+{
+	checkResults(PQprepare(m_conn, name.c_str(), sql.c_str(), 0, NULL));
+}
+
+PGresult *Database_PostgreSQL::checkResults(PGresult *result, bool clear)
+{
+	ExecStatusType statusType = PQresultStatus(result);
+
+	switch (statusType) {
+		case PGRES_COMMAND_OK:
+		case PGRES_TUPLES_OK:
+			break;
+		case PGRES_FATAL_ERROR:
+		default:
+			throw DatabaseException(std::string("PostgreSQL database error: ") + PQresultErrorMessage(result));
+	}
+
+	if (clear)
+		PQclear(result);
+
+	return result;
+}
+
+void Database_PostgreSQL::createDatabase()
+{
+	PGresult *result = checkResults(PQexec(m_conn,
+		"SELECT relname FROM pg_class WHERE relname='blocks';"),
+		false);
+
+	// If table doesn't exist, create it
+	if (!PQntuples(result)) {
+		static const char* dbcreate_sql = "CREATE TABLE blocks ("
+			"posX INT NOT NULL,"
+			"posY INT NOT NULL,"
+			"posZ INT NOT NULL,"
+			"block_data BYTEA,"
+			"PRIMARY KEY (posX,posY,posZ)"
+		");";
+		checkResults(PQexec(m_conn, dbcreate_sql));
+	}
+
+	PQclear(result);
+
+	infostream << "PostgreSQL: Game Database was inited." << std::endl;
+}
+
+
+void Database_PostgreSQL::beginSave()
+{
+	verifyDatabase();
+	checkResults(PQexec(m_conn, "BEGIN;"));
+}
+
+void Database_PostgreSQL::endSave()
+{
+	checkResults(PQexec(m_conn, "COMMIT;"));
+}
+
+bool Database_PostgreSQL::saveBlock(const v3s16 &pos, const std::string &data)
+{
+	verifyDatabase();
+
+	const char *values[] = { (char*)&pos.X, (char*)&pos.Y, (char*)&pos.Z, data.c_str() };
+	const int valLen[] = { sizeof(int), sizeof(int), sizeof(int), (int)data.size() };
+	const int valFormats[] = { 1, 1, 1, 1 };
+
+	execPrepared("write_block", 4, values, valLen, valFormats);
+	return true;
+}
+
+void Database_PostgreSQL::loadBlock(const v3s16 &pos, std::string *block)
+{
+	verifyDatabase();
+
+	const char *values[] = { (char*)&pos.X, (char*)&pos.Y, (char*)&pos.Z };
+	const int valLen[] = { sizeof(int), sizeof(int), sizeof(int) };
+	const int valFormats[] = { 1, 1, 1 };
```

Comment:
No, no, no.  I don't get what it is you changed here but this is all wrong.
pos.X and so on are not integers, and they do not have the same size as integers.  This is an out-of-bounds memory access bug.